lib/spack/spack/build_environment.py

```diff
@@ -835,154 +831,256 @@ def setup_package(pkg, dirty, context="build"):
-        bin_dir = os.path.join(prefix, dirname)
-        if os.path.isdir(bin_dir):
-            env.prepend_path("PATH", bin_dir)
+        if context == "build":
+            # Drop direct run deps in build context
+            # We don't really distinguish between install and build time test deps,
+            # so we include them here as build-time test deps.
+            self.root_deptypes = ["build", "test", "link"]
+        elif context == "test":
+            # This is more of an extended run environment
+            self.root_deptypes = ["test", "run", "link"]
+        elif context == "run":
+            self.root_deptypes = ["run", "link"]
+        else:
+            raise ValueError(f"Unknown context {context}. Should be one of build, test, run")
 
+    def neighbors(self, item):
+        spec = item.edge.spec
+        if spec.dag_hash() in self.root_hashes:
+            deptype = self.root_deptypes
+        else:
+            deptype = ("link", "run")
+        return traverse.sort_edges(spec.edges_to_dependencies(deptype=deptype))
 
-def modifications_from_dependencies(
-    spec, context, custom_mods_only=True, set_package_py_globals=True
-):
-    """Returns the environment modifications that are required by
-    the dependencies of a spec and also applies modifications
-    to this spec's package at module scope, if need be.
 
-    Environment modifications include:
+class Mode(Flag):
+    # Entrypoint spec (a spec to be built; an env root, etc)
+    ROOT = auto()
 
-    - Updating PATH so that executables can be found
-    - Updating CMAKE_PREFIX_PATH and PKG_CONFIG_PATH so that their respective
-      tools can find Spack-built dependencies
-    - Running custom package environment modifications
+    # A spec used at runtime, but no executables in PATH
+    RUNTIME = auto()
+
+    # A spec used at runtime, with executables in PATH
+    RUNTIME_EXECUTABLE = auto()
+
+    # A spec that's a direct build or test dep
+    BUILDTIME_DIRECT = auto()
+
+    # A spec that should be visible in search paths in a build env.
+    BUILDTIME = auto()
+
+    # Flag is set when the (node, mode) is finalized
+    ADDED = auto()
+
+
+def effective_deptypes(specs: list, context="build"):
+    """
+    Given a list of input specs and a context, return a list of tuples of
+    all specs that contribute to (environment) modifications, together with
+    a flag specifying in what way they do so. The list is ordered topologically
+    from root to leaf, meaning that environment modifications should be applied
+    in reverse so that dependents override dependencies, not the other way around.
+    """
+    assert context in ("build", "run", "test")
+
+    visitor = traverse.TopoVisitor(
+        EnvironmentVisitor(specs, context), key=lambda x: x.dag_hash(), root=True, all_edges=True
+    )
+    traverse.traverse_depth_first_with_visitor(traverse.with_artificial_edges(specs), visitor)
+
+    modes = defaultdict(lambda: Mode(0))
```

Comment:
(Minor) I'm in doubt whether a comment to say that this represents the zeroed flag is worth it or not. 🤔

Reply:
done